HACKING.rst

```diff
@@ -138,3 +138,33 @@ Cloud Config Modules
 
 * Any new modules should use underscores in any new config options and not
   hyphens (e.g. `new_option` and *not* `new-option`).
+
+Unit Testing
+------------
+
+cloud-init uses `pytest`_ to run its tests, and has tests written both
+as ``unittest.TestCase`` sub-classes and as un-subclassed pytest tests.
+The following guidelines should be following:
+
+* For ease of organisation and greater accessibility for developers not
+  familiar with pytest, all cloud-init unit tests must be contained
+  within test classes
+
+  * Put another way, module-level test functions should not be used
+
+* pytest test classes should use `pytest fixtures`_ to share
+  functionality instead of inheritance
+* As all tests are contained within classes, it is acceptable to mix
+  ``TestCase`` test classes and pytest test classes within the same
+  test file
+
+  * These can be easily distinguished by their definition: pytest
+    classes will not use inheritance at all, whereas ``TestCase``
+    classes will subclass (indirectly) from ``TestCase``
+
+* pytest tests should use bare ``assert`` statements, to take advantage
```

Comment:
Since we are prescribing testing style, can we also prescribe order of asserts as "<expected_val> == <dynamic_tested_val>" just so we align on testing style, and ease of triage during failed tests.

Reply:
Done.